Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/GeometricFillTool.java

```diff
@@ -176,36 +198,91 @@ public void onRestoreInstanceState(Bundle bundle) {
 
 		BaseShape baseShape = (BaseShape) bundle.getSerializable(BUNDLE_BASE_SHAPE);
 		ShapeDrawType shapeDrawType = (ShapeDrawType) bundle.getSerializable(BUNDLE_SHAPE_DRAW_TYPE);
+		int shapeOutlineWidth = (int) bundle.getSerializable(BUNDLE_OUTLINE_WIDTH);
 
 		if (baseShape != null && shapeDrawType != null
 				&& (this.baseShape != baseShape || this.shapeDrawType != shapeDrawType)) {
 			this.baseShape = baseShape;
 			this.shapeDrawType = shapeDrawType;
+			this.shapeOutlineWidth = shapeOutlineWidth;
 
 			shapeToolOptionsListener.setShapeActivated(baseShape);
+			shapeToolOptionsListener.setDrawTypeActivated(shapeDrawType);
+			shapeToolOptionsListener.setShapeOutlineWidth(shapeOutlineWidth);
 			createAndSetBitmap();
 		}
 	}
 
-	private void drawShape(Canvas drawCanvas, RectF shapeRect, Paint drawPaint, int drawableId) {
-		Rect rect = new Rect((int) shapeRect.left, (int) shapeRect.top, (int) shapeRect.right, (int) shapeRect.bottom);
+	@Override
+	public void draw(Canvas canvas) {
+		super.draw(canvas);
+		createAndSetBitmap();
```

Comment:
Does this mean you are creating a new bitmap on each draw call? Because that's bad for performance. You should not create objects during a drawing call, bitmaps are expensive.